File: src/test/java/com/atlassian/migration/datacenter/api/aws/AWSConfigureEndpointTest.java
/*
 * Copyright 2020 Atlassian
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.atlassian.migration.datacenter.api.aws;

import com.atlassian.migration.datacenter.core.aws.cloud.AWSConfigurationService;
import com.atlassian.migration.datacenter.core.exceptions.InvalidMigrationStageError;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import javax.ws.rs.core.Response;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.Mockito.doThrow;
import static org.mockito.Mockito.verify;

@ExtendWith(MockitoExtension.class)
class AWSConfigureEndpointTest {

    @Mock
    AWSConfigurationService configurationService;

    @InjectMocks
    AWSConfigureEndpoint sut;

    @Test
    void shouldConfigureAWS() throws InvalidMigrationStageError {
        AWSConfigureEndpoint.AWSConfigureWebObject payload = new AWSConfigureEndpoint.AWSConfigureWebObject();
        final String accessKeyId = "accessKeyId";
        final String secretKey = "secretKey";
        final String region = "us-east-1";
        payload.setAccessKeyId(accessKeyId);
        payload.setSecretAccessKey(secretKey);
        payload.setRegion(region);

        Response response = sut.storeAWSCredentials(payload);

        verify(configurationService).configureCloudProvider(accessKeyId, secretKey, region);
        assertEquals(Response.Status.NO_CONTENT.getStatusCode(), response.getStatus());
    }

    @Test
    void shouldReturnFailedWhenInvalidMigrationStage() throws InvalidMigrationStageError {
        AWSConfigureEndpoint.AWSConfigureWebObject payload = new AWSConfigureEndpoint.AWSConfigureWebObject();
        payload.setAccessKeyId("accessKeyId");
        payload.setSecretAccessKey("secretKey");
        payload.setRegion("us-east-1");

        doThrow(new InvalidMigrationStageError("")).when(configurationService).configureCloudProvider(anyString(), anyString(), anyString());

        Response response = sut.storeAWSCredentials(payload);

        assertEquals(Response.Status.CONFLICT.getStatusCode(), response.getStatus());
    }

}

File: src/test/kotlin/com/atlassian/migration/datacenter/api/aws/AWSConfigureEndpointTest.kt
/*
 * Copyright 2020 Atlassian
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.atlassian.migration.datacenter.api.aws

import com.atlassian.migration.datacenter.api.aws.AWSConfigureEndpoint.AWSConfigureWebObject
import com.atlassian.migration.datacenter.core.aws.cloud.AWSConfigurationService
import com.atlassian.migration.datacenter.core.exceptions.InvalidMigrationStageError
import io.mockk.MockKAnnotations
import io.mockk.confirmVerified
import io.mockk.every
import io.mockk.impl.annotations.InjectMockKs
import io.mockk.impl.annotations.MockK
import io.mockk.junit5.MockKExtension
import io.mockk.verify
import javax.ws.rs.core.Response
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith

@ExtendWith(MockKExtension::class)
internal class AWSConfigureEndpointTest {
    @MockK(relaxUnitFun = true)
    lateinit var configurationService: AWSConfigurationService

    @InjectMockKs
    lateinit var sut: AWSConfigureEndpoint

    @BeforeEach
    fun init() {
        MockKAnnotations.init(this)
    }

    @Test
    fun shouldConfigureAWS() {
        val payload = AWSConfigureWebObject()
        val accessKeyId = "accessKeyId"
        val secretKey = "secretKey"
        val region = "us-east-1"
        payload.accessKeyId = accessKeyId
        payload.secretAccessKey = secretKey
        payload.region = region

        val response = sut.storeAWSCredentials(payload)

        verify { configurationService.configureCloudProvider(accessKeyId, secretKey, region) }
        confirmVerified(configurationService)
        assertEquals(Response.Status.NO_CONTENT.statusCode, response.status)
    }

    @Test
    fun shouldReturnFailedWhenInvalidMigrationStage() {
        val payload = AWSConfigureWebObject()
        payload.accessKeyId = "accessKeyId"
        payload.secretAccessKey = "secretKey"
        payload.region = "us-east-1"

        every { configurationService.configureCloudProvider(any(), any(), any()) } throws InvalidMigrationStageError("")

        val response = sut.storeAWSCredentials(payload)
        assertEquals(Response.Status.CONFLICT.statusCode, response.status)
    }
}